pkg/csr/csr.go

```diff
@@ -0,0 +1,236 @@
+	if !hasExactUsages(req, kubeletServerUsages) {
+		return false, errors.Errorf("CSR does not contain required usages: %s", req.Name)
+	}
+
+	var hasOrg bool
+	for i := range csr.Subject.Organization {
+		if csr.Subject.Organization[i] == nodeGroup {
+			hasOrg = true
+			break
+		}
+	}
+	if !hasOrg {
+		return false, errors.Errorf("CSR does not contain required subject organization: %s", req.Name)
+	}
+	return true, nil
+}
+
+// Approve approves a CSR by updating it's status conditions to true if it is a valid CSR
+func (c *CSRApprover) Approve(csr *certificatesv1.CertificateSigningRequest) error {
+
+	if validated, err := c.validateCSRContents(csr); !validated {
+		c.log.Info("CSR contents are invalid for approval by WMCO CSR Approver", "CSR", csr.Name)
+		return errors.Wrapf(err, "could not validate CSR %s contents for approval", csr.Name)
+	}
+
+	csr.Status.Conditions = append(csr.Status.Conditions, certificatesv1.CertificateSigningRequestCondition{
+		Type:           certificatesv1.CertificateApproved,
+		Status:         "True",
+		Message:        "This CSR was approved by the WMCO CSR Approver",
+		LastUpdateTime: metav1.Now(),
+		Reason:         "WMCOApproved",
+	})
+
+	if _, err := c.k8sClientSet.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.Background(),
+		csr.Name, csr, metav1.UpdateOptions{}); err != nil {
+		return errors.Wrapf(err, "could not update conditions for CSR approval %s", csr.Name)
+	}
+	c.log.Info("CSR approved by WMCO CSR Approver", "CSR", csr.Name)
+	return nil
+}
+
+// isNodeClientCert returns true if the CSR is from a  kube-apiserver-client-kubelet signer
+// reference: https://kubernetes.io/docs/reference/access-authn-authz/certificate-signing-requests/#kubernetes-signers
+func isNodeClientCert(csr *certificatesv1.CertificateSigningRequest, x509cr *x509.CertificateRequest) bool {
+	if !reflect.DeepEqual([]string{nodeGroup}, x509cr.Subject.Organization) {
+		return false
+	}
+	if (len(x509cr.DNSNames) > 0) || (len(x509cr.EmailAddresses) > 0) || (len(x509cr.IPAddresses) > 0) {
+		return false
+	}
+	// Check usages include: digital signature, key encipherment and client auth
+	if !hasExactUsages(csr, kubeletClientUsages) {
+		return false
+	}
+	if !strings.HasPrefix(x509cr.Subject.CommonName, NodeUserPrefix) {
+		return false
+	}
+	return true
+}
+
+// hasExactUsages verifies if the required usages exist in the CSR spec
+func hasExactUsages(csr *certificatesv1.CertificateSigningRequest, usages []certificatesv1.KeyUsage) bool {
+	if len(usages) != len(csr.Spec.Usages) {
+		return false
+	}
+
+	usageMap := map[certificatesv1.KeyUsage]struct{}{}
+	for _, u := range usages {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, u := range csr.Spec.Usages {
+		if _, ok := usageMap[u]; !ok {
+			return false
+		}
+	}
+
+	return true
+}
+
+// getDnsAddresses returns list of dns addresses corresponding to the given IPv4 address
+func getDnsAddresses(ipAddr string) []string {
+	addr, _ := net.LookupAddr(ipAddr)
```

Comment:
I should have caught this in my previous review. You are doing a reverse look  up here. What if you are in an environment where this does not resolve? Moreover every IP needn't have a DNS name associated with it. We should discuss this. I would like to understand why this is needed.

Reply:
modified this to checking the platform type and deciding to get the actual hostname vs lookup the hostname depending on the usecase.